bindings/serialize.cc

```diff
@@ -0,0 +1,290 @@
+
+std::vector<int64_t> Sample::getValue() { return value; }
+
+std::vector<Label> Sample::getLabel() { return label; }
+
+void Sample::encode(std::vector<char>* buffer) const {
+  encodeUint64s(1, locationID, buffer);
+  encodeInt64s(2, value, buffer);
+  encodeRepeatedMessage<Label>(3, label, buffer);
+}
+
+Profile::Profile(std::string periodType, std::string periodUnit, int64_t period,
+                 int64_t timeNanos, int64_t durationNanos,
+                 std::string dropFrames, std::string keepFramesX)
+    : period(period),
+      timeNanos(timeNanos),
+      durationNanos(durationNanos),
+      defaultSampleTypeX(0) {
+  // first index of strings must be ""
+  stringID("");
+  this->periodType = ValueType(stringID(periodType), stringID(periodUnit));
+  this->dropFramesX = stringID(dropFrames);
+  this->keepFramesX = stringID(keepFramesX);
+}
+
+void Profile::addSampleType(std::string type, std::string unit) {
+  int64_t typeX = stringID(type);
+  int64_t unitX = stringID(unit);
+  sampleType.push_back(ValueType(typeX, unitX));
+}
+
+void Profile::addSample(std::unique_ptr<Node>& node,
+                        std::deque<uint64_t>* stack) {
+  uint64_t loc = locationID(node);
+  stack->push_front(loc);
+  std::vector<Sample> nodeSamples = node->samples(*stack, this);
+  sample.insert(sample.end(), nodeSamples.begin(), nodeSamples.end());
+}
+
+uint64_t Profile::locationID(std::unique_ptr<Node>& node) {
+  LocationKey key(node->getFileID(), node->lineNumber(), node->columnNumber(),
+                  node->name());
+  auto ids = locationIDMap.find(key);
+  if (ids != locationIDMap.end()) {
+    return ids->second;
+  }
+  uint64_t id = location.size() + 1;
+  std::vector<Line> lines;
+  lines.push_back(line(node));
+  ProfileLocation l = ProfileLocation(id, 0, 0, lines, false);
+  location.push_back(l);
+  locationIDMap.insert(locationIDMap.begin(),
+                       std::pair<LocationKey, int64_t>(key, id));
+  return id;
+}
+
+Line Profile::line(std::unique_ptr<Node>& node) {
+  return Line(functionID(node), node->lineNumber());
+}
+
+int64_t Profile::functionID(std::unique_ptr<Node>& node) {
+  std::string name = node->name();
+  FunctionKey key(node->getFileID(), name);
+  auto ids = functionIDMap.find(key);
+  if (ids != functionIDMap.end()) {
+    return ids->second;
+  }
+  int64_t nameX = stringID(name);
+  int64_t filenameX = stringID(node->filename());
+  int64_t id = function.size() + 1;
+  ProfileFunction f =
+      ProfileFunction(id, nameX, nameX, filenameX, node->lineNumber());
+  function.push_back(f);
+  functionIDMap.insert(functionIDMap.begin(),
+                       std::pair<FunctionKey, int64_t>(key, id));
+  return id;
+}
+
+int64_t Profile::stringID(std::string s) {
+  auto pair = stringIDMap.find(s);
+  if (pair != stringIDMap.end()) {
+    return pair->second;
+  }
+  int64_t id = strings.size();
+  stringIDMap.insert(stringIDMap.begin(),
```

Comment:
This can be just `stringIDMap.insert({s, id});`. First, you used the form of insert that takes hint position as the first argument - http://www.cplusplus.com/reference/map/map/insert/. But this is not useful since you don't have any hint anyway. Second, use shortcuts like {first, second} here to create types that can be deduced by the compiler.

Reply:
Similarly in other places where you insert things.

Reply:
Done.